### PR TITLE
[SYCL-MLIR] Move alloca to the beginning of function

### PR DESCRIPTION
**Pull request by whitneywhtsang**
`Basic/linear-sub_group.cpp` fails when not all `AllocaOp`s are in the beginning of functions.
We moved `memref::AllocaOp`s to the beginning of caller functions in Inliner pass, but it only works for `AllocaOp` in `memref` dialect and only when the problem is exposed from Inliner pass.

This PR removed the inliner code that moves alloca operations, and adds a new utility function to move `LLVM::AllocaOp` right before emitting LLVM IR. AFAIK there are no limitations in LLVM IR that require alloca operation to be at the top of a function. This is just a limitation of the LLVM to SPIRV translator. 

This is therefore a workaround that should be removed once the translator is able to tollerate alloca operation in the middle of a function.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp [hunk 1]
```diff
@@ -828,12 +828,48 @@ forwardPassthroughAttributes(Location loc, std::optional<ArrayAttr> attributes,
   return success();
 }
 
+// This function moves all AllocaOp (and its operands) to the beginning of the
+// function.
+static bool moveAllocaToTop(LLVMFuncOp func) {
```
Comment (etiotto) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:833]: minor: consider returning void

**Review by etiotto — COMMENTED** [review]
Can you add some lit tests to verify that several alloca are injected at the top of the function with their operands in the correct order. 

Also is it possible to call a verifier on the module to ensure the code generated is "correct"

**Comment by whitneywhtsang** [comment]
> Also is it possible to call a verifier on the module to ensure the code generated is "correct"

`verifyModule()` is called at the end of `translateModuleToLLVMIR()`.

**Comment by sommerlukas** [comment]
Should we raise this issue on https://github.com/KhronosGroup/SPIRV-LLVM-Translator, to at least have a discussion with them if this is something they want to address as part of the translator? 

As this is a SPIR-V specific requirement, the translator seems the right place to address this. Downside would of course be that a mere round-trip through the translator would change the position of `alloca`s. 

In the meantime, makes sense to fix this here to keep moving.

**Review by sommerlukas — APPROVED** [review]
No further comments apart from @etiotto's comments.

**Comment by whitneywhtsang** [comment]
> Should we raise this issue on https://github.com/KhronosGroup/SPIRV-LLVM-Translator, to at least have a discussion with them if this is something they want to address as part of the translator?
> 
> As this is a SPIR-V specific requirement, the translator seems the right place to address this. Downside would of course be that a mere round-trip through the translator would change the position of `alloca`s.
> 
> In the meantime, makes sense to fix this here to keep moving.

I will create an issue on https://github.com/KhronosGroup/SPIRV-LLVM-Translator.

**Comment by etiotto** [comment]
> Should we raise this issue on https://github.com/KhronosGroup/SPIRV-LLVM-Translator, to at least have a discussion with them if this is something they want to address as part of the translator?
> 
> As this is a SPIR-V specific requirement, the translator seems the right place to address this. Downside would of course be that a mere round-trip through the translator would change the position of `alloca`s.
> 
> In the meantime, makes sense to fix this here to keep moving.

I believe @whitneywhtsang is trying to purse this with the translator team. Opening an issue with a reduced test is not a bad idea IMO. We discussed it yesterday offline and we both agree this is a temp. workaround.

File: polygeist/tools/cgeist/Test/Verification/move_alloca_to_top.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// RUN: mlir-translate -mlir-to-llvmir -split-input-file -verify-diagnostics %s | FileCheck %s
+
+// CHECK-LABEL: define void @move_alloca(float* %0) {
+// CHECK-DAG:     %{{.*}} = alloca float, i8 1, align 4
+// CHECK-DAG:     %{{.*}} = alloca float, i8 1, align 4
+// CHECK-NEXT:    %4 = load float, float* %0, align 4
```
Comment (etiotto) [inline polygeist/tools/cgeist/Test/Verification/move_alloca_to_top.cpp:6]:
```suggestion
// CHECK-NEXT:    %{{.*}} = load float, float* %0, align 4
```

**Review by etiotto — APPROVED** [review]

File: polygeist/tools/cgeist/Test/Verification/move_alloca_to_top.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/move_alloca_to_top.cpp:6]: After two alloca operation, the load must be %4.

**Comment by whitneywhtsang** [comment]
> I will create an issue on https://github.com/KhronosGroup/SPIRV-LLVM-Translator.

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1883

**Pull request merged**